### PR TITLE
Resolve #2474: Online Indexing: add reverse order option

### LINKED ISSUE #2474: Online Indexing: add reverse order option
This can be useful when the indexing order matters.

### PR DESCRIPTION
**Pull request by jjezra**

### CODE REVIEW

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerIndexFromIndexTest.java [hunk 1]
```diff
@@ -392,8 +431,9 @@ void testIndexFromIndexWithDuplicates() {
         assertEquals(0, timer.getCount(FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RECORDS_INDEXED));
     }
 
-    @Test
-    void testIndexFromIndexPersistentContinuation() {
+    @ParameterizedTest
+    @ValueSource(ints = {0, 1, 2, 3})
+    void testIndexFromIndexPersistentContinuation(int reverseSeed) {
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerIndexFromIndexTest.java:436]: This should call `assertAllValidated`

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java [hunk 2]
```diff
@@ -256,8 +265,9 @@ void testMultiTargetMismatchStateFailure() {
         assertAllValidated(indexes);
     }
 
-    @Test
-    void testMultiTargetPartlyBuildFailure() {
+    @ParameterizedTest
+    @ValueSource(ints = {0, 1, 2, 3, 4, 5, 6, 7})
+    void testMultiTargetPartlyBuildFailure(int reverseSeed) {
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java:270]: This test should call `assertAllValidated`

**Review by ScottDugas — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java [hunk 3]
```diff
@@ -834,6 +840,57 @@ private <T> CompletableFuture<Boolean> iterateCursorOnly(@Nonnull FDBRecordStore
                 });
     }
 
+    @SuppressWarnings("squid:S00107") // too many parameters
+    private <T> CompletableFuture<Boolean> iterateCursorOnlyReverse(@Nonnull FDBRecordStore store,
+                                                                    @Nonnull RecordCursorResult<T> rangeCursor,
```
[the remark this answers is not in the document]
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java:845]: ✅

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java [hunk 4]
```diff
@@ -834,6 +840,57 @@ private <T> CompletableFuture<Boolean> iterateCursorOnly(@Nonnull FDBRecordStore
...
+                    if (null == rec) {
+                        return AsyncUtil.READY_TRUE; // next
+                    }
+                    // This record should be indexed. Add it to the transaction.
+                    if (isIdempotent) {
+                        store.addRecordReadConflict(rec.getPrimaryKey());
+                    }
+                    timerIncrement(FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RECORDS_INDEXED);
+
+                    final CompletableFuture<Void> updateMaintainer = updateMaintainerBuilder(store, rec);
```
[the remark this answers is not in the document]
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java:880]: Yes. It is built for multi target, so it updates each target index that doesn't filter out this record type.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerIndexFromIndexTest.java [hunk 5]
```diff
@@ -121,7 +124,8 @@ void testIndexFromIndexSimple() {
         assertEquals(numRecords, timer.getCount(FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RECORDS_INDEXED));
     }
 
-    @Test
+    @ParameterizedTest
+    @BooleanSource
```
[the remark this answers is not in the document]
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerIndexFromIndexTest.java:128]: ✅

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java [hunk 6]
```diff
@@ -163,15 +163,25 @@ private CompletableFuture<Boolean> buildRangeOnly(@Nonnull FDBRecordStore store,
...
         });
     }
 
+    private CompletableFuture<Boolean> postIterateRangeOnly(IndexingRangeSet rangeSet, boolean hasMore,
+                                                            AtomicReference<RecordCursorResult<FDBIndexedRecord<Message>>> lastResult,
+                                                            Tuple rangeStart, Tuple rangeEnd, boolean isReverse) {
+        if (isReverse) {
+            Tuple cont = hasMore ? lastResult.get().get().getIndexEntry().getKey() : rangeStart;
+            return rangeSet.insertRangeAsync(packOrNull(cont), packOrNull(rangeEnd), true)
+                    .thenApply(ignore -> hasMore || rangeStart != null);
```
[the remark this answers is not in the document]
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java:177]: ✅

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java [hunk 7]
```diff
@@ -163,15 +163,25 @@ private CompletableFuture<Boolean> buildRangeOnly(@Nonnull FDBRecordStore store,
...
+        } else {
+            Tuple cont = hasMore ? lastResult.get().get().getIndexEntry().getKey() : rangeEnd;
+            return rangeSet.insertRangeAsync(packOrNull(rangeStart), packOrNull(cont), true)
+                    .thenApply(ignore -> hasMore || rangeEnd != null);
```
[the remark this answers is not in the document]
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java:181]: ✅

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java [hunk 8]
```diff
@@ -185,14 +185,25 @@ private CompletableFuture<Boolean> buildRangeOnly(@Nonnull FDBRecordStore store,
...
+                                                            AtomicReference<RecordCursorResult<FDBStoredRecord<Message>>> lastResult,
+                                                            Tuple rangeStart, Tuple rangeEnd, boolean isReverse) {
+        if (isReverse) {
+            Tuple cont = hasMore ? lastResult.get().get().getPrimaryKey() : rangeStart;
+            return insertRanges(targetRangeSets, packOrNull(cont), packOrNull(rangeEnd))
+                    .thenApply(ignore -> hasMore || rangeStart != null);
+        } else {
+            Tuple cont = hasMore ? lastResult.get().get().getPrimaryKey() : rangeEnd;
+            return insertRanges(targetRangeSets, packOrNull(rangeStart), packOrNull(cont))
+                    .thenApply(ignore -> hasMore || rangeEnd != null);
```
[the remark this answers is not in the document]
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java:203]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerIndexFromIndexTest.java (hunk 1 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerIndexFromIndexTest.java:436]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java [hunk 9]
```diff
@@ -144,8 +147,9 @@ void testMultiTargetSimple() {
         assertAllValidated(indexes);
     }
 
-    @Test
-    void testMultiTargetContinuation() {
+    @ParameterizedTest
+    @BooleanSource
+    void testMultiTargetContinuation(boolean reverseScan) {
```
[the remark this answers is not in the document]
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java:152]: There are non-idempotent (COUT) indexes in the tests, do you expect SUM to behave differently?

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java (hunk 2 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java:270]: This test asserts typeStamp failures. But I check that `assertAllValidated` is added to the tests that successfully finish indexing.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java:152]: To be on the safe side, I've added SUM indexes to the tests.

**Review by ScottDugas — APPROVED** [review]

**Pull request merged**